Sources/StreamChat/AudioPlayer/AudioPlaying.swift

```diff
@@ -0,0 +1,375 @@
+            guard
+                let self = self,
+                self.context.isSeeking == false
+            else {
+                return
+            }
+
+            self.updateContext { value in
+                let currentTime = player.currentTime().seconds
+                value.currentTime = currentTime.isFinite && !currentTime.isNaN
+                    ? TimeInterval(currentTime)
+                    : .zero
+
+                value.isSeeking = false
+
+                value.state = player.rate != 0 ? .playing : .paused
+
+                value.rate = .init(rawValue: player.rate)
+            }
+        }
+
+        playerObserver.addTimeControlStatusObserver(
+            player
+        ) { [weak self] newValue in
+            guard
+                let self = self,
+                let newValue = newValue
+            else {
+                return
+            }
+
+            let currentPlaybackState = self.context.state
+
+            self.updateContext { value in
+                switch (newValue, currentPlaybackState) {
+                case (.playing, .playing), (.paused, .paused):
+                    break
+                case (.paused, .playing), (.paused, .loading):
+                    value.state = .paused
+                    value.rate = .zero
+                case (.playing, .paused), (.playing, .stopped), (.playing, .loading):
+                    value.state = .playing
+                default:
+                    log.debug("\(type(of: self)): No action for transition \(currentPlaybackState) -> \(newValue)")
+                }
+            }
+        }
+
+        playerObserver.addStoppedPlaybackObserver(
+            queue: nil
+        ) { [weak self] playerItem in
+            guard
+                let self = self,
+                let playerItemURL = (playerItem.asset as? AVURLAsset)?.url,
+                let currentItemURL = (player.currentItem?.asset as? AVURLAsset)?.url,
+                playerItemURL == currentItemURL
+            else {
+                return
+            }
+            var newContext = self.context
+            newContext.state = .stopped
+            newContext.currentTime = 0
+            newContext.rate = .zero
+            newContext.isSeeking = false
+            self.context = newContext
+
+            self.delegate?.audioPlayer(self, didUpdateContext: self.context)
+        }
+    }
+
+    /// Provides thread-safe updates for the player's context and makes sure to forward any updates
+    /// to the the delegate
+    private func updateContext(
+        _ newContextProvider: (inout AudioPlaybackContext) -> Void
+    ) {
+        var newContext = context
+        newContextProvider(&newContext)
+        context = newContext
+        delegate?.audioPlayer(self, didUpdateContext: context)
+    }
+
+    /// It's used by the assetPropertyLoader to handle the completion (successful or failed) of duration's
+    /// asynchronous loading.
+    private func handleDurationLoading(
```

Comment:
Feels a bit unnatural that this method is defined before its usage (file placement wise)